dbms/src/Interpreters/IdentifierSemantic.cpp

```diff
@@ -5,6 +5,61 @@
 namespace DB
 {
 
+namespace ErrorCodes
+{
+    extern const int AMBIGUOUS_COLUMN_NAME;
+}
+
+namespace
+{
+
+const DatabaseAndTableWithAlias & extractTable(const DatabaseAndTableWithAlias & table)
+{
+    return table;
+}
+
+const DatabaseAndTableWithAlias & extractTable(const TableWithColumnNames & table)
+{
+    return table.first;
+}
+
+template <typename T>
+IdentifierSemantic::ColumnMatch tryChooseTable(const ASTIdentifier & identifier, const std::vector<T> & tables,
+                                               size_t & best_table_pos, bool allow_ambiguous)
+{
+    using ColumnMatch = IdentifierSemantic::ColumnMatch;
+
+    best_table_pos = 0;
+    auto best_match = ColumnMatch::NoMatch;
+    size_t same_match = 0;
+
+    for (size_t i = 0; i < tables.size(); ++i)
+    {
+        auto match = IdentifierSemantic::canReferColumnToTable(identifier, extractTable(tables[i]));
+        if (value(match))
```

Comment:
So NoMatch is implicitly assigned 0, then we static_cast it to UInt32 using a special function, and then implicitly convert to bool? I'd just write `match != NoMatch`.